numpy/_core/src/multiarray/mapping.c

```diff
@@ -3013,6 +3013,8 @@ PyArray_MapIterNew(npy_index_info *indices , int index_num, int index_type,
         if (extra_op == NULL) {
             goto fail;
         }
+        // extra_op_dtype might have been replaced, so get a new reference
+        extra_op_dtype = PyArray_DESCR(extra_op);
```

Comment:
LGTM, but the ref-counting looks off.  We own a reference to `extra_op_dtype` but incref it above for cleanup.

I would be OK with just using `Py_SETREF(extra_op_dtype, Py_NewRef(...))` though.  That way we just always own one reference.